### PR TITLE
Plugin service tests

### PR DESCRIPTION
**Pull request by charliescheer**
### Description

This PR adds unit tests to the parts of  WordPressKit-iOS that handle plugins.  The updates cover these files:
PluginDirectoryFeedPage.swift
PluginState.swift
SitePluginCapabilities.swift
SitePlugin.swift
PluginDirectoryEntry.swift
PluginServiceRemote.swift
PluginDirectoryServiceRemote.swift

### Testing Details
PR only deals with the addition of new tests, to test it, simply run the testing suite WordPressKitTests.  All tests should pass.

### CODE REVIEW

File: WordPressKit/PluginDirectoryEntry.swift [hunk 1]
```diff
@@ -12,10 +12,10 @@ public struct PluginDirectoryEntry {
     public let author: String
     public let authorURL: URL?
 
-    private let descriptionHTML: String?
-    private let installationHTML: String?
-    private let faqHTML: String?
-    private let changelogHTML: String?
+    internal let descriptionHTML: String?
```
Comment (jkmassel) [inline WordPressKit/PluginDirectoryEntry.swift]: `internal` should be the default, so it's not strictly necessary to have them here, but it doesn't hurt anything either :)

File: WordPressKit/PluginServiceRemote.swift [hunk 2]
```diff
@@ -16,7 +16,7 @@ public class PluginServiceRemote: ServiceRemoteWordPressComREST {
                 failure(ResponseError.decodingFailure)
                 return
             }
-
+
```
Comment (jkmassel) [inline WordPressKit/PluginServiceRemote.swift]: Might need to set Xcode to remove empty whitespace in lines

File: WordPressKitTests/Mock Data/plugin-activate-malformed-response.json [hunk 3]
```diff
@@ -0,0 +1,18 @@
+{
+    "slug" "code-snippets",
```
Comment (jkmassel) [inline WordPressKitTests/Mock Data/plugin-activate-malformed-response.json]: This response could probably be just this one line if you're just testing for invalid JSON

Comment (jkmassel) [reply WordPressKitTests/Mock Data/plugin-activate-malformed-response.json]: Applies to others as well!

File: WordPressKitTests/MockPluginDirectoryEntryProvider.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
...g stats, calendar, and author widgets</li>\n<li>[paid] High-speed, ad-free, and high-definition video hosting</li>\n</ul>\n<h4>Discussion and Community</h4>\n<p>Create a connection with your readers and keep them coming back to your site with:</p>\n<ul>\n<li>[free] Email subscriptions</li>\n<li>[free] Comment login with Facebook, Twitter, and Google</li>\n<li>[free] Fully-customizable contact forms</li>\n<li>[free] Infinite scroll for your posts</li>\n</ul>\n<h4>Expert Support</h4>\n<p>We have an entire team of Happiness Engineers ready to help you. Ask your questions in the support forum, or <a href=\"https://jetpack.com/contact-support\" rel=\"nofollow\">contact us directly</a>.</p>\n<h4>Paid Services</h4>\n<p>Most of Jetpack&#8217;s features and services are free. Jetpack also provides advanced security and backup services, video hosting, site monetization, priority support, and more SEO tools in three <a href=\"https://jetpack.com/pricing?from=wporg\" rel=\"nofollow\">simple and affordable plans</a>.</p>\n<h4>Get Started</h4>\n<p>Installation is free, quick, and easy. Set up <a href=\"https://jetpack.com/install?from=wporg\" rel=\"nofollow\">the free plan</a> in minutes.</p>\n"
```
Comment (jkmassel) [inline WordPressKitTests/MockPluginDirectoryEntryProvider.swift]: WDYT about putting this in a JSON file and parsing it in? It's a lot of boilerplate IMHO? Not necessarily wrong...tbh I'm not sure if that's best either but something I was thinking about.

File: WordPressKitTests/MockPluginStateProvider.swift [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    """.utf8)
+        
+        return updateStateJSON
+    }
+    
+    static func getEncodedUpdateState(state: PluginState.UpdateState) -> Data {
+        var data = Data()
+        let encoder = JSONEncoder()
+        
+        do {
```
Comment (jkmassel) [inline WordPressKitTests/MockPluginStateProvider.swift]: Because this is a test helper, it might make sense to just `throw` out of this method instead of handling it – if it's not working, that failure should probably bubble up to the individual test.

File: WordPressKitTests/PluginDirectoryTests.swift [hunk 6]
```diff
@@ -77,6 +77,28 @@ class PluginDirectoryTests: XCTestCase {
             XCTFail(error.localizedDescription)
         }
     }
+    
+    func testValidatePluginDirectoryFeedResponseSucceeds() {
```
Comment (jkmassel) [inline WordPressKitTests/PluginDirectoryTests.swift]: These tests can `throw` on their own, and if they do it'll be counted as a failure – no need to catch it! (this is a relatively new feature that may require a newer Xcode version – if so, we can update it in CI)

File: WordPressKitTests/PluginDirectoryTests.swift [hunk 7]
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
...
+            let slugs = response.plugins.map { $0.slug }
+            XCTAssertEqual(response.pageMetadata.pluginSlugs, slugs)
+            
+        } catch {
+            XCTFail("Failed decoding plugin \(error)")
+        }
+    }
+    
+    func testPluginFeedPageDirectoryEquatable() {
+        let jetpackMockPath = Bundle(for: type(of: self)).path(forResource: "plugin-directory-jetpack", ofType: "json")!
```
Comment (jkmassel) [inline WordPressKitTests/PluginDirectoryTests.swift]: We repeat this three-line pattern a lot – perhaps it could be extracted into a helper?

File: WordPressKitTests/PluginDirectoryTests.swift [hunk 8]
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
...
+        let jetpackPluginMockPath = Bundle(for: type(of: self)).path(forResource: "plugin-directory-jetpack", ofType: "json")!
+        let json = JSONLoader().loadFile(jetpackPluginMockPath) as AnyObject
+        guard let response = json as? [String : AnyObject] else {
+            return
+        }
+
+        do {
+            let directoryEntry = try PluginDirectoryEntry(responseObject: response)
+
+            XCTAssertEqual(directoryEntry.name, "Jetpack by WordPress.com")
```
Comment (jkmassel) [inline WordPressKitTests/PluginDirectoryTests.swift:265]: This is good, but it comes back to what I mentioned about dynamic naming – if you can get this dynamically generated (ideally using non-ascii characters, HTML entities, whatever other edge cases you can think of), it'll result in a better test because it stressed the code rather than just validating inputs to outputs.

File: WordPressKitTests/PluginDirectoryTests.swift [hunk 9]
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
...
+            let data = try encoder.encode(plugin)
+        } catch {
+            XCTFail("Convert to JSON Failed")
+        }
+    }
+
+    func testPluginDirectoryFeedTypeSlugReturn() {
+        let pluginDirectoryFeedTypeNewest = PluginDirectoryFeedType.newest
+        let pluginDirectoryFeedTypePopular = PluginDirectoryFeedType.popular
+        let pluginDirectoryFeedTypeSearch = PluginDirectoryFeedType.search(term: "blocks")
```
Comment (jkmassel) [inline WordPressKitTests/PluginDirectoryTests.swift]: Nitpick – these lines may not be necessary – you could probably use `PluginDirectoryFeedType.newest.slug` in the assertion?

File: WordPressKitTests/PluginServiceRemoteTests.swift [hunk 10]
```diff
@@ -65,4 +84,438 @@ class PluginServiceRemoteTests: RemoteTestCase, RESTTestable {
...
+            let expected = PluginServiceRemote.ResponseError.decodingFailure as NSError
+            XCTAssertEqual(error, expected)
+            
+            expect.fulfill()
+        }
+        
+        waitForExpectations(timeout: timeout, handler: nil)
+    }
+    
+    func testPluginIDEncoding() {
```
Comment (jkmassel) [inline WordPressKitTests/PluginServiceRemoteTests.swift:188]: This is a great type of test to have

**Review by jkmassel — APPROVED** [review]
This looks very good – I've left a few little comments, but overall I think this is a good set of tests to validate this area of the project!

File: WordPressKitTests/PluginDirectoryTests.swift [hunk 11]
```diff
@@ -143,4 +165,166 @@ class PluginDirectoryTests: XCTestCase {
...
+
+            let slugs = response.plugins.map { $0.slug }
+            XCTAssertEqual(response.pageMetadata.pluginSlugs, slugs)
+            
+        } catch {
+            XCTFail("Failed decoding plugin \(error)")
+        }
+    }
+    
+    func testPluginFeedPageDirectoryEquatable() {
```
Comment (jkmassel) [inline WordPressKitTests/PluginDirectoryTests.swift:177]: One other thing I'd suggest adding tests for – are there any feed page directories that superficially might look the same, but actually aren't? Is that something that can be tested?

File: WordPressKit/PluginServiceRemote.swift (hunk 2 above)
Comment (charliescheer) [reply WordPressKit/PluginServiceRemote.swift]: Done, didn't know that was an option :)
Thanks!
